docs/design/proposals/rbd-qos.md

```diff
@@ -0,0 +1,269 @@
+   * Retrieve the QOS configuration from the volumeContext in NodePublishRequest
+   * Identify the rbd device using the NodeStageVolumePath
+   * Get the pod UUID from the NodeStageVolume
+   * Set io.max file in all the containers in the pod
+
+#### Drawbacks of StorageClass QOS
+
+1. No way to update the QOS at runtime
+1. Need to take a backup and restore to New QOS StorageClass
+1. Delete and Recreate the PV object
+
+### Add support for VolumeAttributeClass
+
+```yaml
+apiVersion: storage.k8s.io/v1alpha1
+kind: VolumeAttributesClass
+metadata:
+  name: silver
+parameters:
+  MaxReadIOOperationsPerSecond: ""
+  MaxWriteIOOperationsPerSecond: ""
+  MaxReadBytesPerSecond: ""
+  MaxWriteBytesPerSecond: ""
+```
+
+VolumeAttributesClassName is a new parameter in the PVC object the user can
+choose from and this can also be updated or removed later.
+
+This new VolumeAttributeClass is designed to keep storage that supports setting
+QOS at the storage level which means setting some configuration at the storage
+(like QOS for nbd)
+
+#### Implementation of VolumeAttributeClass QOS
+
+1. Modify CSIDriver object to pass pod details to the NodePublish Volume RPC
+1. Add support in cephCSI to expose ModifyVolume RPC
+1. CephCSI will store QOS in the rbd image metadata
+1. During NodeStage operation retrieve the image metadata and store it in
+   stagingPath
+1. Whenever a new pod comes in apply the QOS
+
+#### Drawbacks of VolumeAttributeClass QOS
+
+One problem with above is all application need to be scaled downed and scaled
+up to get the new QOS value even though its changed in the PVC object, this is
+sometime impossible as it will have downtime.
+
+### VolumeAttributeClass with NodePublish Secret
+
+1. Modify CSIDriver object to pass pod details to the NodePublish Volume
+1. Add support in cephCSI to expose ModifyVolumeRPC call
+1. CephCSI will store QOS in the rbd image metadata
+1. During NodePublish operation retrieve the QOS from image metadata
+1. Whenever a new pod comes in apply the QOS
+
+This solution addresses the aforementioned issue, but it requires a secret to
+communicate with the ceph cluster. Therefore, we must create a new
+PublishSecret for the storageClass, which may be beneficial when Kubernetes
+eventually enables Node operations.
+
+Both options 2 and 3 are contingent upon changes to the CSI spec and Kubernetes
+support. Additionally, VolumeAttributeClass is currently being developed within
+the Kubernetes realm and will initially be in the Alpha stage. Consequently, it
+will be disabled by default, but it does offer the following advantages.
+
+1. No Restore/Clone operation is required to change the QOS
+1. Easily QOS can be changed for existing PVC only with second approach not
+   with third as it needs new secret.
+
+#### Hybrid Approach
+
+Considering the advantages and drawbacks, we can use StorageClass and
+VolumeAttributeClass to support QoS, with VolumeAttributeClass taking
+precedence over StorageClass.
```

Comment:
@nixpanic not as of today as kubernetes is working on it, if not we can look for adding it in future

Reply:
> @nixpanic this is deferred for Kubernetes 1.29 [kubernetes/enhancements#3751 (comment)](https://github.com/kubernetes/enhancements/issues/3751#issuecomment-1640690606) and am not sure it will happen (hope it will happen), and as it goes with Alpha, Beta Feature cycle it takes more time for users to use this in production and not everyone will agree to upgrade to latest kubernetes for this one. I think it better to have both solutions and IMO it is not a maintenance and testing burden for this one as its minimal one, later once VolumeAttributeClass is stable we can deprecate the support for storageclass Qos.

+1